File: ag-compiler/src/scanner.rs
use super::{token::*, token_type::*};
use std::ops::Range;

macro_rules! str {
    () => {
        String::new()
    };
    ($slice: literal) => {
        ($slice).to_string()
    };
    (true, $expr: expr) => {
        format!("{}", ($expr))
    };
    (false, $expr: expr) => {
        format!("{:?}", ($expr))
    };
}

macro_rules! ty {
    ($kind: ident) => {
        TokenType::$kind
    };
}

macro_rules! r#match {
    ($scan: ident, $char: expr) => {
        ($scan).r#match(($char))?
    };
}

macro_rules! tok {
    ($scan: expr, $type: ident) => {
        ($scan).tok(ty!($type))
    };
}
macro_rules! ok_tok {
    ($scan: expr, $type: ident) => {
        Ok(tok!($scan, $type))
    };
}

/// The scanner's error type. It consists of a message, the line and the column it occurred on,
/// in the order (message, line, column).
pub(super) type ScanError = (String, usize, usize);
type ScanResult = Result<Token, ScanError>;

/// A scanner that scans through source code and parses it into easy-to-use `Token`s for the compiler.
/// It works with a shared string reference to conserve memory, and uses `Vec<String>` for characters
/// because a `Vec<char>` is inaccurate due to Unicode, and using `String` allows us to fully parse a character
/// thanks to the [unicode_segmentation][https://crates.io/crates/unicode-segmentation] crate.
///
/// It uses a small counter, `current`, to peek past the
pub struct Scanner<'src> {
    source: &'src String,
    chars: Vec<char>,
    start: usize,
    current: usize,
    line: usize,
    columns: Range<usize>,
    last: Option<TokenType>,
}
impl<'src> Scanner<'src> {
    pub fn new<'s: 'src>(source: &'s String) -> Self {
        Self {
            source,
            chars: source.chars().collect(),
            start: 0,
            current: 0,
            line: 1,
            columns: 0..0,
            last: None,
        }
    }
    /// Reinitialize the scanner for new source code.
    pub fn re_init<'s: 'src>(&mut self, source: &'s String) {
        self.source = source;
        self.chars = source.chars().collect();
        self.current = 0;
        self.line = 1;
        self.columns = 0..0;
        self.last = None;
    }
    fn is_at_end(&self) -> bool {
        self.current > self.chars.len()
        // ^^^ because we use `[self.current - 1]`
    }
    fn current(&self) -> &char {
        &self.chars[self.current - 1]
    }
    fn advance(&mut self) -> Result<&char, ScanError> {
        self.current += 1;
        self.columns.end += 1;
        if self.is_at_end() {
            Err((
                "Unexpected end of file".to_string(),
                self.line,
                self.columns.end,
            ))
        } else {
            Ok(self.current())
        }
    }
    fn peek(&self) -> &char {
        &self.chars[self.current]
    }
    fn peek_next(&self) -> Option<&char> {
        if self.chars.len() <= self.current + 1 {
            None
        } else {
            Some(&self.chars[self.current + 1])
        }
    }
    fn tok(&self, kind: TokenType) -> Token {
        Token::new(
            kind,
            self.source[self.start..self.current].to_string(),
            self.line,
            self.columns.start..self.columns.end,
        )
    }
    fn r#match(&mut self, kind: char) -> Result<bool, ScanError> {
        Ok(if self.peek() == &kind {
            self.advance()?;
            true
        } else {
            false
        })
    }
    pub fn scan(&mut self) -> ScanResult {
        self.start = self.current;
        let char = match self.advance() {
            Ok(c) => c,
            Err(_) => return Ok(Token::eof(self.line, self.columns.start..self.columns.end)),
        };

        match char {
            // groups
            '(' => ok_tok!(self, LeftParen),
            ')' => ok_tok!(self, RightParen),
            '{' => ok_tok!(self, LeftBrace),
            '}' => ok_tok!(self, RightBrace),
            '[' => ok_tok!(self, LeftBracket),
            ']' => ok_tok!(self, RightBracket),
            // logical
            '=' => {
                if r#match!(self, '=') {
                    ok_tok!(self, EqualEqual)
                } else if r#match!(self, '>') {
                    ok_tok!(self, FatArrow)
                } else {
                    ok_tok!(self, Equal)
                }
            }
            '!' => {
                if r#match!(self, '=') {
                    ok_tok!(self, BangEqual)
                } else {
                    ok_tok!(self, Bang)
                }
            }
            // comparison
            '<' => {
                if r#match!(self, '=') {
                    ok_tok!(self, LessEqual)
                } else if r#match!(self, ':') {
                    ok_tok!(self, Inherits)
                } else {
                    ok_tok!(self, Less)
                }
            }
            '>' => {
                if r#match!(self, '=') {
                    ok_tok!(self, GreaterEqual)
                } else {
                    ok_tok!(self, Greater)
                }
            }
            // arithmetic
            '+' => {
                if r#match!(self, '=') {
                    ok_tok!(self, PlusEqual)
                } else {
                    ok_tok!(self, Plus)
                }
            }
            '-' => {
                if r#match!(self, '=') {
                    ok_tok!(self, MinusEqual)
                } else if r#match!(self, '>') {
                    ok_tok!(self, RetArrow)
                } else {
                    ok_tok!(self, Minus)
                }
            }
            '*' => {
                if r#match!(self, '*') {
                    if r#match!(self, '=') {
                        ok_tok!(self, PowerEqual)
                    } else {
                        ok_tok!(self, Power)
                    }
                } else if r#match!(self, '=') {
                    ok_tok!(self, StarEqual)
                } else {
                    ok_tok!(self, Star)
                }
            }
            '/' => {
                if r#match!(self, '=') {
                    ok_tok!(self, SlashEqual)
                } else {
                    ok_tok!(self, Slash)
                }
            }
            '%' => {
                if r#match!(self, '=') {
                    ok_tok!(self, PercentEqual)
                } else {
                    ok_tok!(self, Percent)
                }
            }
            // bitwise
            '&' => {
                if r#match!(self, '=') {
                    ok_tok!(self, BitAndEqual)
                } else {
                    ok_tok!(self, BitAnd)
                }
            }
            '|' => {
                if r#match!(self, '=') {
                    ok_tok!(self, BitOrEqual)
                } else if r#match!(self, '>') {
                    ok_tok!(self, Pipe)
                } else {
                    ok_tok!(self, BitOr)
                }
            }
            '^' => {
                if r#match!(self, '=') {
                    ok_tok!(self, BitXorEqual)
                } else {
                    ok_tok!(self, BitXor)
                }
            }
            // other
            ';' => ok_tok!(self, Semicolon),
            '\n' => ok_tok!(self, Newline),
            ':' => ok_tok!(self, Colon),
            '.' => {
                if r#match!(self, '.') {
                    if r#match!(self, '=') {
                        ok_tok!(self, InclusiveRange)
                    } else {
                        ok_tok!(self, ExclusiveRange)
                    }
                } else {
                    ok_tok!(self, Dot)
                }
            }
            // literals
            '"' => self.string(),
            '\'' => self.char(),
            c => {
                if c.is_alphabetic() {
                    self.ident()
                } else if c.is_numeric() {
                    self.number()
                } else {
                    Err((
                        format!("Character {c} is not recognized"),
                        self.line,
                        self.columns.end,
                    ))
                }
            }
        }
    }

    fn number(&mut self) -> ScanResult {
        while self.peek().is_numeric() {
            self.advance()?;
        }
        if self.peek() == &'.' && self.peek_next() == Some(&'1') {
            self.advance()?;

            while self.peek().is_numeric() {
                self.advance()?;
            }

            ok_tok!(self, Decimal)
        } else {
            ok_tok!(self, Integer)
        }
    }
    fn ident(&mut self) -> ScanResult {
        while self.peek().is_alphanumeric() {
            self.advance()?;
        }
        let slice = &self.source[self.start..self.current];
        return match slice {
            "and" => ok_tok!(self, And),
            "or" => ok_tok!(self, Or),
            "let" => ok_tok!(self, Let),
            "var" => ok_tok!(self, Var),
            "global" => ok_tok!(self, Global),
            "mut" => ok_tok!(self, Mut),
            "if" => ok_tok!(self, If),
            "unless" => ok_tok!(self, Unless),
            "while" => ok_tok!(self, While),
            "until" => ok_tok!(self, Until),
            "for" => ok_tok!(self, For),
            "in" => ok_tok!(self, In),
            "loop" => ok_tok!(self, Loop),
            "func" => ok_tok!(self, Func),
            "Self" => ok_tok!(self, CapitalSelf),
            "self" => ok_tok!(self, _Self),
            "class" => ok_tok!(self, Class),
            "pub" => ok_tok!(self, Pub),
            "true" => ok_tok!(self, True),
            "false" => ok_tok!(self, False),
            _ => ok_tok!(self, Identifier),
        };
    }
    fn char(&mut self) -> ScanResult {
        // we do not give it a length now, as we need to handle unicode graphemes correctly
        while !r#match!(self, '\'') {
            self.advance()?;
        }
        self.advance()?;

        ok_tok!(self, Character)
    }
    fn string(&mut self) -> ScanResult {
        while !r#match!(self, '"') {
            self.advance()?;
        }
        self.advance()?; // consume end quote

        ok_tok!(self, String)
    }
}

File: AgateCS/Scanner/Scanner.cs
using System.Text;
using AgateCS.Scanner.ScanUtils;
using AgateCS.Utils;
using static AgateCS.Scanner.TokenType;

namespace AgateCS.Scanner
{
    namespace ScanUtils
    {
        public interface IScanResult
        {
            public string Message();
        }

        public class InvalidChar(char c, string reason) : IScanResult
        {
            public string Reason = reason;
            public readonly char Character = c;

            public override string ToString()
            {
                return $"{Reason} character '{Character}'";
            }
            public string Message()
            {
                return ToString();
            }
        }

        public readonly struct UnexpectedEnd(string msg) : IScanResult
        {
            public readonly string Msg = msg;

            public string Message()
            {
                return Msg;
            }
        }
    }
    
    public static class Scanner
    {
        private static string Source = "";

        public static void ReInit(string source)
        {
            Source = source;
            Line = 1;
            Start = 0;
            Current = 0;
            Columns.Start = 0;
            Columns.End = 0;
        }

        private static int Line = 1;
        private static int Start = 0;
        private static readonly Range<int> Columns = new(0, 0);
        private static int Current = 0;

        private static void SkipWhiteSpace()
        {
            while (true)
            {
                switch (Source[Current])
                {
                    case ' ':
                    case '\r':
                    case '\t':
                        break;
                    case '#':
                        while (!IsAtEnd())
                        {
                            Advance();
                            if (Peek() == '\n')
                            {
                                break;
                            }
                        }
                        break;
                    default:
                        return;
                }
            }
        }
        private static char MoveForward()
        {
            Columns.End++;
            return Source[Current++];
        }
        private static char Advance() => IsAtEnd() ? '\0' : MoveForward();
        private static char Peek(int count = 0) => (Current + 1 + count >= Source.Length) ? '\0' : Source[Current + 1 + count];
        private static bool Match(char c)
        {
            if (Peek() == c)
            {
                Advance();
                return true;
            }
            return false;
        }
        private static Token MakeToken(TokenType type) => new(Columns, new(Line, Line), type, Source[Start..Current]);
        private static Token MakeToken(TokenType baseType, char match, TokenType matchType) => MakeToken(Match(match) ? matchType : baseType);
        private static bool IsAtEnd() => Current >= Source.Length;

        public static IScanResult ScanToken()
        {
            SkipWhiteSpace();

            Start = Current;
            Columns.Start = Columns.End;

            char c = Advance();

            IScanResult result =  c switch
            {
                '(' => MakeToken(LeftParen),
                ')' => MakeToken(RightParen),
                '{' => MakeToken(LeftBrace),
                '}' => MakeToken(RightBrace),
                '[' => MakeToken(LeftBracket),
                ']' => MakeToken(RightBracket),
                '!' => MakeToken(Bang, '=', BangEqual),
                '=' => Match('>') ? MakeToken(FatArrow) : MakeToken(Equal, '=', EqualEqual),
                '+' => MakeToken(Plus, '=', PlusEqual),
                '-' => Match('>') ? MakeToken(RetArrow) : MakeToken(Minus, '=', MinusEqual),
                '*' => Match('*') ? MakeToken(Power, '*', PowerEqual) : MakeToken(Star, '=', StarEqual),
                '/' => MakeToken(Slash, '=', SlashEqual),
                '%' => MakeToken(Percent, '=', PercentEqual),
                '&' => MakeToken(BitAnd, '=', BitAndEqual),
                '|' => Match('>') ? MakeToken(Pipe) : MakeToken(BitOr, '=', BitOrEqual),
                '^' => MakeToken(BitXor, '=', BitXorEqual),
                '>' => MakeToken(Greater, '=', GreaterEqual),
                '<' => Match(':') ? MakeToken(Inherits) : MakeToken(Less, '=', LessEqual),
                '\n' => NewLine(),
                ';' => MakeToken(Semicolon),
                '?' => MakeToken(Question),
                ':' => Match('=') ? MakeToken(ColonEqual) : MakeToken(Colon, ':', ColonColon),
                '.' => Match('.') ? MakeToken(ExclusiveRange, '=', InclusiveRange) : MakeToken(Dot),
                '"' => String(),
                '\'' => Char(),


                _ when char.IsDigit(c) => Number(),
                _ when char.IsLetter(c) || c == '_' => Ident(),
                _ => new InvalidChar(c, "Unkown"),
            };

            if (AgateCS.Config.DebugScan && result is Token token)
            {
                Console.WriteLine($"|- Debug.Scan: Produced token {token.Message()}");
            }
            return result;
        }

        private static Token NewLine()
        {
            Line++;
            Columns.End = 0;
            return MakeToken(Newline);
        }
        private static IScanResult Char()
        {
            if (IsAtEnd() || Peek() == '\'' || Peek() == '\n')
            {
                return new UnexpectedEnd("Unended or invalid character literal");
            }

            char c = Advance();
            
            if (c == '\\')
            {
                switch(Advance())
                {
                    case 'n':
                        c = '\n';
                        break;
                    case 't':
                        c = '\t';
                        break;
                    case 'r':
                        c = '\r';
                        break;
                    case '\\':
                        c = '\\';
                        break;
                    case '\'':
                        c = '\'';
                        break;
                    case '0':
                        c = '\0';
                        break;
                    default:
                        return new InvalidChar(c, "Invalid escape");

                }
            }

            if (Peek() != '\'')
            {
                return new UnexpectedEnd("Character literal is too long");
            }

            Advance();

            return new Token(Columns, new(Line, Line), Character, Source[Start..Current], c.ToString());
        }
        private static IScanResult String()
        {
            int StartLine = Line;
            StringBuilder builder = new();
            while (Peek() != '"' && !IsAtEnd())
            {
                char c = Advance();

                if (c == '\\')
                {
                    char next = Advance();
                    builder.Append(next switch
                    {
                        'n' => '\n',
                        't' => '\t',
                        'r' => '\r',
                        '\\' => '\\',
                        '"' => '"',
                        '\'' => '\'',
                        '0' => '\0',
                        _ => next,
                    });
                }
                else
                {
                    builder.Append(c);
                }
            }

            if (IsAtEnd()) return new UnexpectedEnd("Unended string literal");
            else Advance();

            return new Token(Columns, new(StartLine, Line), StringLit, Source[Start..Current], builder.ToString());
        }

        private static Token Ident()
        {
            while ((char.IsDigit(Peek()) || char.IsLetter(Peek()) || Peek() == '_') && !IsAtEnd())
            {
                Advance();
            }

            string text = Source[Start..Current];
            return text switch
            {
                "and" => MakeToken(And),
                "or" => MakeToken(Or),
                "let" => MakeToken(Let),
                "var" => MakeToken(Var),
                "global" => MakeToken(Global),
                "mut" => MakeToken(Mut),
                "if" => MakeToken(If),
                "unless" => MakeToken(Unless),
                "while" => MakeToken(While),
                "until" => MakeToken(Until),
                "loop" => MakeToken(Loop),
                "for" => MakeToken(For),
                "in" => MakeToken(In),
                "func" => MakeToken(Func),
                "class" => MakeToken(Class),
                "self" => MakeToken(Self),
                "Self" => MakeToken(CapitalSelf),
                "mod" => MakeToken(Mod),
                "pub" => MakeToken(Pub),
                "true" => MakeToken(True),
                "false" => MakeToken(False),
                "null" => MakeToken(Null),
                "is" => MakeToken(Is),
                _ => MakeToken(Identifier),
            };
        }
        private static Token Number()
        {
            while (char.IsDigit(Peek()) && !IsAtEnd())
            {
                Advance();
            }
            if (Peek() == '.' && char.IsDigit(Peek(1)))
            {
                Advance();
                while (char.IsDigit(Peek()) && !IsAtEnd())
                {
                    Advance();
                }
                return MakeToken(FloatLit);
            }
            else
            {
                return MakeToken(IntegerLit);
            }
        }
    }
}
